### PR TITLE
Add API functionality to beatmap favourite button

### PR DESCRIPTION
**Pull request by EVAST9919**

### CODE REVIEW

File: osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs [hunk 1]
```diff
@@ -67,16 +53,22 @@ private void load()
...
+
+            Action = () =>
+            {
+                if (loading.State.Value == Visibility.Visible)
+                    return;
+
+                loading.Show();
+
+                request?.Cancel();
+                request = new PostBeatmapFavouriteRequest(BeatmapSet.Value?.OnlineBeatmapSetID ?? 0, favourited.Value ? BeatmapFavouriteAction.UnFavourite : BeatmapFavouriteAction.Favourite);
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: This probably shouldn't send a request at all if `BeatmapSet.Value` is null.

File: osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs [hunk 2]
```diff
@@ -67,16 +53,22 @@ private void load()
...
+                request.Success += () => favourited.Value = !favourited.Value;
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: I'm a bit worried about this potentially desyncing with the actual web state. Does that endpoint not actually return whether or not the favourite was added or removed? I'd be less concerned about potential discrepancies then.

**Review by bdach — COMMENTED** [review]
Looks to be working ok and I *think* it's race-free - I have some nitpicks though.

File: osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs (hunk 2 above)
Comment (EVAST9919) [reply osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: afaik it returns total favourite beatmaps amount

Comment (bdach) [reply osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: Agh, that doesn't help here much. I guess this is passable then.

Comment (iiSaLMaN) [reply osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: I agree with @bdach, you should open an issue web-side for this, this may be dangerous.

Comment (peppy) [reply osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: Yep, should ask for API change.

**Comment by bdach** [comment]
* Tagging as blocked pending resolution of ppy/osu-web#5108.
* By the way, another thing to check here is how the favourite button handles hitting the maximum allowed number of favourited beatmaps.

**Comment by bdach** [comment]
Unblocking after close of web issue, I guess the toggle should be *mostly* safe as-is.

**Comment by EVAST9919** [comment]
How about just send another request after the first one to redownload beatmap metadata and see, whether this beatmap is favourited?

**Comment by smoogipoo** [comment]
You would need to constantly send requests if you wanted to keep it somewhat consistent. Seems not worthwhile.

**Comment by EVAST9919** [comment]
Okay, after hitting a limit of favourited beatmaps there's no actual API response. It worth it to create a web issue.

**Comment by nanaya** [comment]
That's not what that endpoint is for. Use `beatmapsets/:id` to get its latest data.

**Comment by peppy** [comment]
To clarify, the endpoint *will* return an error if you are at the limit of allowed favourites.

File: osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs [hunk 3]
```diff
@@ -68,6 +69,18 @@ private void load(IAPIProvider api)
                 request?.Cancel();
                 request = new PostBeatmapFavouriteRequest(BeatmapSet.Value?.OnlineBeatmapSetID ?? 0, favourited.Value ? BeatmapFavouriteAction.UnFavourite : BeatmapFavouriteAction.Favourite);
                 request.Success += () => favourited.Value = !favourited.Value;
+                request.Failure += exception =>
+                {
+                    if (exception.Message == "UnprocessableEntity")
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: This feels wrong due to hinging on a magic string (which *is* the name of an HTTP error code, but still), and [I don't think 422 isn't guaranteed to be returned on this error only either](https://github.com/ppy/osu-web/blob/666a8135dc576afffb185dd5315d95fb1c4ca302/app/helpers.php#L560-L563). If anything is to be displayed to the user here, it's probably whatever web returns in the `error` message.

Comment (peppy) [reply osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: osu-web should probably be returning an error in the body.

Comment (peppy) [reply osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: surprise, it already was

**Comment by peppy** [comment]
@smoogipoo requesting a quick check on this, as i made many changes.

can split out `APIRequest `error handling and `DirectOverlay` fixes if you'd prefer (they are relatively minor, though).

**Review by peppy — DISMISSED** [review]

File: osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs [hunk 4]
```diff
@@ -55,28 +42,49 @@ private void load()
...
+                    favourited.Toggle();
+                    loading.Hide();
+                };
+
+                request.Failure += e =>
                 {
-                    pink.FadeOut(200);
-                    icon.Icon = FontAwesome.Regular.Heart;
-                }
+                    notifications.Post(new SimpleNotification
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: This leads to `TestSceneBeatmapSetOverlay.TestAvailability()` crashing the test browser - clicking the button in that test raises a 404, which then causes an attempt to access `notifications` which is null inside the test. This was a problem before too. Unsure if we just want to null-check this, refactor somehow or leave as-is.

File: osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs [hunk 5]
```diff
@@ -55,28 +42,49 @@ private void load()
...
+            }, true);
+
+            Action = () =>
+            {
+                if (loading.State.Value == Visibility.Visible)
                     return;
 
-                favourited.Value = setInfo.NewValue.OnlineInfo.HasFavourited;
-            });
+                // guaranteed by disabled state abvove.
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: I feel horrible for even suggesting this, but
```suggestion
                // guaranteed by disabled state above.
```

**Review by bdach — COMMENTED** [review]
Not @smoogipoo, but took a look. The new changes are good with me. I got some things to point out but they're not supremely relevant.

File: osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs (hunk 4 above)
Comment (iiSaLMaN) [reply osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs]: should null check as null notifications overlay is allowed here.

**Comment by smoogipoo** [comment]
The button needs to disable when API is logged out.

**Comment by smoogipoo** [comment]
It also needs to have a test. I highly suggest moving the fix out into its own separate PR.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**